lib/compiler.js

```diff
@@ -0,0 +1,129 @@
+'use strict';
+
+/**
+ * See spec: https://www.w3.org/TR/webvtt1/#file-structure
+ */
+
+function CompilerError (message, error) {
+  this.message = message;
+  this.error = error;
+}
+CompilerError.prototype = Object.create(Error.prototype);
+
+function compile (input) {
+
+  if (!input) {
+    throw new CompilerError('Input must be non-null');
+  }
+  if (typeof input !== 'object') {
+    throw new CompilerError('Input must be an object');
+  }
+  if (input.isArray) {
+    throw new CompilerError('Input cannot be array');
+  }
+
+  if (!input.valid) {
+    throw new CompilerError('Input must be valid');
+  }
+
+
+  let output = 'WEBVTT\n';
+
+  input.cues.forEach((cue) => {
+    output += '\n';
+    output += compileCue(cue);
+    output += '\n';
+  });
+
+  return output;
+}
+
+/**
+ * Compile a single cue block.
+ *
+ * @param {array} cue Array of content for the cue
+ *
+ * @returns {object} cue Cue object with start, end, text and styles.
+ *                       Null if it's a note
+ */
+function compileCue (cue) {
+  // TODO: check for malformed JSON
+  if (typeof cue !== 'object') {
+    throw new CompilerError('Cue malformed: not of type object');
+  }
+  if (typeof cue.identifier !== 'string' &&
+    typeof cue.identifier !== 'number' &&
+    cue.identifier !== null) {
+    throw new CompilerError(`Cue malformed: identifier value is not a string.
+    ${JSON.stringify(cue)}`);
+  }
+  if (isNaN(cue.start)) {
+    throw new CompilerError(`Cue malformed: null start value.
+    ${JSON.stringify(cue)}`);
+  }
+  if (isNaN(cue.end)) {
+    throw new CompilerError(`Cue malformed: null end value.
+    ${JSON.stringify(cue)}`);
+  }
+  if (cue.start >= cue.end) {
+    throw new CompilerError(`Cue malformed: start timestamp greater than end
+    ${JSON.stringify(cue)}`);
+  }
+  if (typeof cue.text !== 'string') {
+    throw new CompilerError(`Cue malformed: null text value.
+    ${JSON.stringify(cue)}`);
+  }
+  if (typeof cue.styles !== 'string') {
+    throw new CompilerError(`Cue malformed: null styles value.
+    ${JSON.stringify(cue)}`);
+  }
+
+  let output = '';
+  if (cue.identifier.length > 0) {
+    output += `${cue.identifier}\n`;
+  }
+
+  const startTimestamp = convertTimestamp(cue.start);
+  const endTimestamp = convertTimestamp(cue.end);
+
+  output += `${startTimestamp} --> ${endTimestamp} ${cue.styles}\n${cue.text}`;
```

Comment:
The space between `${endTimestamp} ${cue.styles}` causes a redundant trailing space if there are no `styles`, it should check if `cue.styles` is set and then adding it + a space